cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,831 @@
+ * @{
+ */
+
+/**
+ * @brief Retrieve a canonical index associated with a given memory type.
+ *
+ * For variants based on memory type, this index can be used to help keep a
+ * consistent ordering of the memory types in the variant.
+ */
+inline auto constexpr variant_index_from_memory_type(raft::memory_type mem_type)
+{
+  return static_cast<std::underlying_type_t<raft::memory_type>>(mem_type);
+}
+
+/**
+ * @brief Retrieve a type from a variant based on a given memory type.
+ */
+template <raft::memory_type MemType, typename Variant>
+using alternate_from_mem_type =
+  std::variant_alternative_t<variant_index_from_memory_type(MemType) % std::variant_size_v<Variant>,
+                             Variant>;
+
+/**
+ * @brief A variant of container policies for each memory type which can be
+ * used to build the default container policy for a buffer.
+ */
+template <typename T>
+using default_container_policy_variant = std::variant<host_vector_policy<T>,
+                                                      device_uvector_policy<T>,
+                                                      managed_uvector_policy<T>,
+                                                      pinned_vector_policy<T>>;
+
+/**
+ * @brief A template used to translate a variant of underlying mdarray
+ * container policies into a container policy that can be used by an mdbuffer.
+ */
+template <typename ElementType,
+          typename ContainerPolicyVariant = default_container_policy_variant<ElementType>>
+struct default_buffer_container_policy {
+  using element_type = ElementType;
+  using value_type   = std::remove_cv_t<element_type>;
+
+  using container_policy_variant =
+    std::variant<host_device_accessor<std::variant_alternative_t<0, ContainerPolicyVariant>, static_cast<memory_type>(0)>, host_device_accessor<std::variant_alternative_t<1, ContainerPolicyVariant>, static_cast<memory_type>(1)>, host_device_accessor<std::variant_alternative_t<2, ContainerPolicyVariant>, static_cast<memory_type>(2)>, host_device_accessor<std::variant_alternative_t<3, ContainerPolicyVariant>, static_cast<memory_type>(3)>, >;
+  template <raft::memory_type MemType>
+  using container_policy = alternate_from_mem_type<MemType, container_policy_variant>;
+
+ private:
+  template <std::size_t index>
+  using container_policy_at_index = std::variant_alternative_t<index, ContainerPolicyVariant>;
+
+ public:
+  using container_type_variant =
+    std::variant<typename container_policy_at_index<0>::container_type,
+                 typename container_policy_at_index<1>::container_type,
+                 typename container_policy_at_index<2>::container_type,
+                 typename container_policy_at_index<3>::container_type>;
+
+  template <raft::memory_type MemType>
+  using container_type = alternate_from_mem_type<MemType, container_type_variant>;
+
+  using accessor_policy_variant =
+    std::variant<host_device_accessor<typename container_policy_at_index<0>::accessor_policy,
```

Comment:
Should we directly be able to use `typename container_policy_at_index<0>::accessor_policy` without the wrap of `host_device_accessor` here?

Reply:
Potentially, but not definitively. If someone passed a custom container policy variant that does not wrap the accessor policy, we want to wrap it here because we need it wrapped for all the subsequent machinery.